### PR TITLE
Change the plugin tool to properly build the dev channel

### PR DESCRIPTION
**Pull request by stevemessick**
The initial dev channel support only built the specs that identified as dev channel specs. This makes a number of changes in how things are built:

- A dev channel build produces all versions
- A master branch build produces all versions, plus the version for integration tests
- A release build produces only the versions intended to be published (no change)

The dev channel version is now derived from the previous release and the number of weeks since its branch was committed. This is how it looks in the plugin management page:

Screen Shot 2020-02-05 at 10 46 46 AM

(There may be unresolved issues with point releases. I couldn't test that.)

I did a little refactoring, too. Initialization of globals was really confusing, and now that it has to be done async it needed to be cleaned up.

@devoncarew

### CODE REVIEW

**Comment by stevemessick** [comment]
Looks like this needs more git magic.

**Comment by devoncarew** [comment]
taking a look at this now

File: tool/plugin/lib/plugin.dart [hunk 1]
```diff
@@ -241,7 +253,7 @@ Future<bool> performReleaseChecks(ProductCommand cmd) async {
           log('the .travis.yml file needs updating: plugin generate');
         }
       } else {
-        log('the current git branch must be named "release_${cmd.releaseMajor}"');
+        log('the current git branch must be named "$expectedName"');
```
Comment (devoncarew) [inline tool/plugin/lib/plugin.dart]: Do we need to fail the build here?

File: tool/plugin/lib/plugin.dart [hunk 2]
```diff
@@ -271,6 +283,15 @@ void separator(String name) {
   log('${ansi.red}${ansi.bold}$name${ansi.none}', indent: false);
 }
 
+int get devBuildNumber {
```
Comment (devoncarew) [inline tool/plugin/lib/plugin.dart]: Do we have the ability to publish a rev to a weekly dev build? What if we find a P0 in the dev channel and what to patch it in order to help validate a fix? a dev channel `.1`

File: tool/plugin/lib/plugin.dart [hunk 3]
```diff
@@ -316,6 +348,12 @@ Future<int> zip(String directory, String outFile) async {
   return await exec('zip', args, cwd: p.dirname(directory));
 }
 
+String _nextRelease() {
+  var num = RegExp(r'release_(\d+)').matchAsPrefix(lastReleaseName).group(1);
```
Comment (devoncarew) [inline tool/plugin/lib/plugin.dart]: `num` is a type; I assume this is valid Dart (as it analyzes correctly), but perhaps name the variable something else?

File: tool/plugin/lib/plugin.dart [hunk 4]
```diff
@@ -632,7 +672,7 @@ class BuildCommand extends ProductCommand {
         processedFile.writeAsStringSync(source);
       }
 
-      if (spec.channel == 'stable') {
+      if (!spec.version.startsWith('4.')) {
```
Comment (devoncarew) [inline tool/plugin/lib/plugin.dart]: Why check against a specific number? If we need to check against `4.` perhaps document in-line for other readers.

File: tool/plugin/lib/plugin.dart (hunk 1 above)
Comment (stevemessick) [reply tool/plugin/lib/plugin.dart]: It does fail. The function returns false, which causes failure.

File: tool/plugin/lib/plugin.dart (hunk 2 above)
Comment (stevemessick) [reply tool/plugin/lib/plugin.dart]: We don't have that currently. We can add it if needed. We'd probably have to add a --patch option or similar, and it should work for stable, master, and dev builds.

File: tool/plugin/lib/plugin.dart (hunk 3 above)
Comment (stevemessick) [reply tool/plugin/lib/plugin.dart]: Good point. Fixed.

File: tool/plugin/lib/plugin.dart (hunk 4 above)
Comment (stevemessick) [reply tool/plugin/lib/plugin.dart]: That's standard practice; checking stable was a mistake. The edits need to be made regardless of which channel we are building. 

There is a TODO upstream. Someday we should refactor the editing code into a separate object.

**Comment by stevemessick** [comment]
Thanks for the review! I captured some work items in https://github.com/flutter/flutter-intellij/issues/4367.

**Pull request merged**